p2p/transport/webrtc/stream.go

```diff
@@ -167,65 +181,134 @@ func (s *stream) Reset() error {
 	return closeReadErr
 }
 
+func (s *stream) closeForShutdown(closeErr error) {
+	defer s.cleanup()
+
+	s.mx.Lock()
+	defer s.mx.Unlock()
+
+	s.closeForShutdownErr = closeErr
+	s.SetReadDeadline(time.Now().Add(-1 * time.Hour))
+	select {
+	case s.sendStateChanged <- struct{}{}:
+	default:
+	}
+}
+
 func (s *stream) SetDeadline(t time.Time) error {
 	_ = s.SetReadDeadline(t)
 	return s.SetWriteDeadline(t)
 }
 
 // processIncomingFlag process the flag on an incoming message
-// It needs to be called with msg.Flag, not msg.GetFlag(),
-// otherwise we'd misinterpret the default value.
 // It needs to be called while the mutex is locked.
 func (s *stream) processIncomingFlag(flag *pb.Message_Flag) {
 	if flag == nil {
 		return
 	}
 
 	switch *flag {
-	case pb.Message_FIN:
-		if s.receiveState == receiveStateReceiving {
-			s.receiveState = receiveStateDataRead
-		}
 	case pb.Message_STOP_SENDING:
-		if s.sendState == sendStateSending {
+		// We must process STOP_SENDING after sending a FIN(sendStateDataSent). Remote peer
+		// may not send a FIN_ACK once it has sent a STOP_SENDING
+		if s.sendState == sendStateSending || s.sendState == sendStateDataSent {
 			s.sendState = sendStateReset
 		}
 		select {
 		case s.sendStateChanged <- struct{}{}:
 		default:
 		}
+	case pb.Message_FIN_ACK:
+		s.sendState = sendStateDataReceived
+		select {
+		case s.sendStateChanged <- struct{}{}:
+		default:
```

Comment:
Future cleanup: make some form of `notifyStateChanged` function.

Reply:
I've replaced the three channels with 1 `writeStateChanged` and using a method to notify on the channel removing all the `select` blocks everywhere.